### PR TITLE
Feature/KAH-97 (Adjust agents service for stateless app)

### PR DESCRIPTION
**Pull request by arispen**
added agentStatusSchema to agentSchema in agent.model

all of the data is persisted in model, except intervalId

as commented in the agents.service:

> // TODO: refactor into scheduler microservice
// why?
// - in node.js, setInterval returns a reference to Timout class instance, not intervalId handle number
//   that's why it's impossible to save a reference to it in a model

### CODE REVIEW

File: server/api/services/map-execution.helper.js [hunk 1]
```diff
@@ -177,69 +210,71 @@ module.exports = {
...
- * @return {Array} - uuid of ancestors
- */
+   * return ancestors for a certain node
+   * @param nodeUuid
+   * @param structure
+   * @return {Array} - uuid of ancestors
+   */
   findAncestors(nodeUuid, structure) {
-    const links = structure.links.filter((o) => o.targetId === nodeUuid);
+    const links = structure.links.filter(o => o.targetId === nodeUuid);
```
Comment (ChBernat) [inline server/api/services/map-execution.helper.js]: What does o stand for?

File: server/api/services/map-execution.helper.js [hunk 2]
```diff
@@ -177,69 +210,71 @@ module.exports = {
...
- * @return {boolean} - false in case another agent got to process first
- */
+   *
+   * @param {*} executionAgents
+   * @param {string} processUUID
+   * @param {*} agentKey
+   * @return {boolean} - false in case another agent got to process first
+   */
   isThisTheFirstAgentToGetToTheProcess(executionAgents, processUUID, agentKey) {
     for (const i in executionAgents) {
```
Comment (ChBernat) [inline server/api/services/map-execution.helper.js]: Oh, come on. Single letter variables are so uncool...

File: server/api/services/map-execution.helper.js [hunk 3]
```diff
@@ -1,155 +1,185 @@
...
 
+async function filterLiveAgents(agents) {
+  const agentsStatus = Object.assign(
+    {},
+    await agentsService.getAllAgentsStatus()
+  );
+  const allAgents = [];
+  agents.forEach(agentObj => {
+    const agentAlive = _.find(agentsStatus, agent => {
+      return agent.id === agentObj.id && agent.alive; // check if this is the condition:
```
[the remark this answers is not in the document]
Comment (arispen) [reply server/api/services/map-execution.helper.js]: check our lint configuration, it shows error when I put parentheses here

File: server/api/services/map-execution.helper.js (hunk 1 above)
Comment (arispen) [reply server/api/services/map-execution.helper.js]: link, i guess

File: server/api/models/agent.model.js [hunk 4]
```diff
@@ -1,24 +1,45 @@
-const mongoose = require('mongoose');
+const mongoose = require("mongoose");
 const Schema = mongoose.Schema;
 
+const agentStatusSchema = new Schema({
+  attributes: [],
+  _id: Schema.Types.ObjectId,
+  name: String,
+  url: String,
+  publicUrl: String,
```
[the remark this answers is not in the document]
Comment (arispen) [reply server/api/models/agent.model.js]: you are right I think it's not used anywhere so I remove it

[the remark this answers is not in the document]
Comment (arispen) [reply server/api/models/agent.model.js]: i will remove the redundancy

File: server/api/services/agents.service.js [hunk 5]
```diff
@@ -1,492 +1,622 @@
...
+  let agentStatus = await getAgentStatus(agent.key);
+  if (!agentStatus || !agentStatus.following) {
+    agentStatus = agent.toJSON();
+    agentStatus.alive = false;
+    agentStatus.following = true;
+
+    await setDefaultUrl(agent);
+    await saveStatusToAgent(agent, agentStatus);
+  }
+  startInterval(agent);
```
Comment (DavidIzaac) [inline server/api/services/agents.service.js]: Why startInterval is at the end of the function and not at the begining ?

Comment (arispen) [reply server/api/services/agents.service.js]: for better predicatbility

**Comment by arispen** [comment]
> In the code tab, the type of currentAgent variable has to be "KaholoAgent". Instead we are getting "any"

not an issue! it's working!

**Comment by arispen** [comment]
> When executing a map with 2 agents, only one of them execute the map.

fixed!

**Comment by arispen** [comment]
finally, after many hours of work and help from @DavidIzaac , we've managed to fix all the features that have broken because of the changes, uff 😌

**Pull request merged**